design_drafts/design_of_controllers_and_their_status.md

```diff
@@ -0,0 +1,17 @@
+# Design proposals for controllers
+
+When writing controllers - there is no consistent naming convention we are using for different variables and this is quite confusing, especially, when debugging things.
+
+Proposal is to have the following names for the 4 important values in controllers:
+
+- **Reference** - input to controller from a topic or preceding controller.
+- **Feedback** - measured state of the system controller controls.
+- **Error** - control error (*Reference* - *Feedback*)
+- **Output** - controller's output, i.e., values written to command interfaces. This are either values of references for the following controller or input to the controlled system.
+
+This is summarized in the following figure:
+
+![Control loop with names](images/control_loop_naming.png)
+
+Beside this issue, use of word **state** is not really suitable for outputting controller's debug information.
+Proposal is to rename this topic to **status** which every controller then **must have** outputting at least the four values mentioned above (if applicable).
```

Comment:
High-level comment: I'm not a native speaker either, but IIUC, the *state* of something in computer science is definitely different from the *status* of something.

In the context of `ros2_control`, I'd expect the status of a controller to be *running* or *starting*, *stopping*, etc.

Its *internal state* would include that status, but `state != status`, as the state of the controller would also include the values of the various member variables (so probably at least the current value for the `reference`, `error`, etc).

Summarising: the name `state` seems apt for what it's supposed to be publishing. Perhaps `internal_` could be prefix to it, but renaming it to `status` would seem like it could be confusing (what status exactly? Of what?).

Reply:
I understand your point. Our perspective was that a `state` is something related to control theory and means a _state variable_.
Thinking of Matlab/Simulink, maybe `signals` is another possibility?

Reply:
Given the fact robotics is inherently multi-disciplinary, I'm not too optimistic it'll be possible to find a name that would be intuitively 'correct' for all future readers/developers.

re: signals: that would again be something that has a 'clear' meaning only in a subfield / limited context.

What about just explaining what sort of information the `state` topic carries? It isn't very complex, and a good enough explanation should clarify things already (should only take a couple of sentences).

It would also not require changing the name of a topic which has been around for a very long time already.

Reply:
@gavanderhoorn I would argue exactly the opposite. “state” is often mean in ROS 2 “Lifecycle state”. And also, the controller usually has much more data, except pure “signals” or “states” in control terms. Typically, there are also certain internal setups of the controller that are relevant to be shown and are useful for debugging.

Reply:
@christophfroehlich do you agree with this new formulation?

Reply:
Fine for me. I would have voted for a different name, but I'm happy to have the values no matter how the topic is named!

Reply:
now even more concrete naming.